homeassistant/components/sia/hub.py

```diff
@@ -0,0 +1,138 @@
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PORT, CONF_PROTOCOL, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, EventOrigin, HomeAssistant
+from homeassistant.helpers import device_registry as dr
+
+from .const import (
+    CONF_ACCOUNT,
+    CONF_ACCOUNTS,
+    CONF_ENCRYPTION_KEY,
+    CONF_IGNORE_TIMESTAMPS,
+    CONF_ZONES,
+    DEFAULT_TIMEBAND,
+    DOMAIN,
+    IGNORED_TIMEBAND,
+    PLATFORMS,
+    SIA_EVENT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SIAHub:
+    """Class for SIA Hubs."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+    ):
+        """Create the SIAHub."""
+        self._hass: HomeAssistant = hass
+        self._entry: ConfigEntry = entry
+        self._port: int = int(entry.data[CONF_PORT])
+        self._title: str = entry.title
+        self._accounts: list[dict[str, Any]] = deepcopy(entry.data[CONF_ACCOUNTS])
+        self._protocol: str = entry.data[CONF_PROTOCOL]
+        self.sia_accounts: list[SIAAccount] | None = None
+        self.sia_client: SIAClient = None
+
+    async def async_setup_hub(self) -> None:
+        """Add a device to the device_registry, register shutdown listener, load reactions."""
+        self.update_accounts()
+        device_registry = await dr.async_get_registry(self._hass)
+        for acc in self._accounts:
+            account = acc[CONF_ACCOUNT]
+            device_registry.async_get_or_create(
+                config_entry_id=self._entry.entry_id,
+                identifiers={(DOMAIN, f"{self._port}_{account}")},
+                name=f"{self._port} - {account}",
+            )
+        self._entry.async_on_unload(
+            self._entry.add_update_listener(self.async_config_entry_updated)
+        )
+        self._entry.async_on_unload(
+            self._hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, self.async_shutdown)
+        )
+
+    async def async_shutdown(self, _: Event = None) -> None:
+        """Shutdown the SIA server."""
+        await self.sia_client.stop()
+
+    async def async_create_and_fire_event(self, event: SIAEvent) -> None:
+        """Create a event on HA's bus, with the data from the SIAEvent.
+
+        The created event is handled by default for only a small subset for each platform (there are about 320 SIA Codes defined, only 22 of those are used in the alarm_control_panel), a user can choose to build other automation or even entities on the same event for SIA codes not handled by the built-in platforms.
+
+        """
+        _LOGGER.debug(
+            "Adding event to bus for code %s for port %s and account %s",
+            event.code,
+            self._port,
+            event.account,
+        )
+        self._hass.bus.async_fire(
+            event_type=SIA_EVENT.format(self._port, event.account),
+            event_data=event.to_dict(encode_json=True),
```

Comment:
Please define the event data keys explicitly so we keep control of them in the integration.

Reply:
I did that initially but it is a list of about 40 fields, so that doesn't do much for the readability, and I want to give users the options to do something with any of them (see the comment above), and it makes the code on the alarm_control_panel side much easier to read and understand because it recreates the object rather then having a dict with a subset of fields.

Reply:
This will become an interface for this integration so we must define it here and not allow it to change unchecked. Please define the keys explicitly.

Reply:
ok

Reply:
for this, does it have to be a flat dict, or can it be a hierarchy? I have some fields that are objects themselves...

Reply:
A value can be a dict, that's ok.